### PR TITLE
Fix Biggoron trade flag conflicts and adult trade item plando bug

### LINKED ISSUE #2061: Biggoron Item is repeatable (or not obtainable at all?)
flagrama — Yesterday at 6:40 PM
https://github.com/OoTRandomizer/OoT-Randomizer/blob/9bbf665578b257ac24f9406bd9a91f4771bba41e/ASM/c/trade_quests.c#L144-L155 if I'm reading this right, if adult trade is off a flag never gets set for claim check being turned in, so turning in the claim check is always valid since the slot doesn't change on trade? and it looks like vanilla depends on bgsFlag being set or not, which I believe only gets set in rando if you find the bgs, so there is no story flag triggered for it? 
(idk much about the new trade stuff, so that's just a guess from a quick repo search though)

### PR DESCRIPTION
**Pull request by mracsys**
Fixes #2061 

The `get_item` check to see if a trade quest item should be marked as traded was being triggered inadvertently by Medigoron. Buying his item before trading in the claim check to Biggoron would mark the claim check as traded and prevent obtaining Biggoron's item due to the dupe prevention scheme.

Additionally, with full adult trade shuffle off, the claim check traded flag was never set, allowing infinite duping. Traded flags in general should not be set with full trade shuffle off because trade items may revert. This is not the case for claim check, so an exception is made to always set its traded flag regardless of trade quest shuffle settings. 

Tested all of the unshuffled adult trade quest starting with pocket cucco and full trade quest shuffle starting with all trade quest items. Medigoron interactions tested only with claim check with and without full trade shuffle.

Validation in-game can be performed watching the custom trade quest flags (u32) at address `0x8011B16C`.

Better documentation of the flags is added to [extra_save_data.md](Notes/extra_save_data.md). I also added explicit listings of each used scene for each feature using the `unk_00_` scene flags, as well as adding Death Mountain Crater to the list of scenes used for scrubs (this caused a previous trade quest bug when it unknowingly had a conflict with the child scrub's scene flag).

Although unrelated, I added the previous Biggoron dupe prevention flag system to the trade quest notes with a brief description for posterity.

Finally, I found a plando bug with adult trade shuffle off and a specific adult trade item plando'd to a location. The internal item selected to be shuffled when building the item pool was not updated when that item was replaced with the plando'd item. The item pool trade item would then be used to pre-set trade flags up to its point in the quest, since those flags enable other checks such as Granny's Potion Shop. If the plando'd item happened to be earlier in the trade sequence than the random item pool trade item, the trade quest became impossible because the plando'd item's traded flag would be set at patch time, triggering dupe prevention in-game. Now, if adult trade shuffle is off and an explicit trade item is plando'd to a location, the plando'd item is set as the internal trade item for flag setting purposes. This bug should not affect the `#AdultTrade` plando alias.

### CODE REVIEW

File: ASM/c/get_items.c [hunk 1]
```diff
@@ -424,8 +424,15 @@ void get_item(z64_actor_t *from_actor, z64_link_t *link, int8_t incoming_item_id
         row = get_item_row(item_id);
         if (row) {
             int16_t action_id = row->action_id;
-            if (CFG_ADULT_TRADE_SHUFFLE && action_id > 0 && from_actor->actor_id != 0x0A && IsAdultTradeItem(action_id)) {
-                if (action_id == Z64_ITEM_BIGGORON_SWORD) {
+            // Set adult trade item "traded" flags to prevent duping.
+            // Only necessary for full adult trade shuffle, except for Biggoron who always sets his flag.
+            if ((CFG_ADULT_TRADE_SHUFFLE || action_id == Z64_ITEM_BIGGORON_SWORD) // full adult trade shuffle on, or vanilla incoming item is Biggoron Sword
```
Comment (mracsys) [inline ASM/c/get_items.c:429]: The comments here are kind of messy. I'm open to clearer ways to write this.

File: Notes/extra-save-data.md [hunk 2]
```diff
@@ -10,17 +10,67 @@ The ammunition count fields for fire arrows (save context + 0x90) and Din's Fire
...
 * Unlocked ocarina notes: Scene 0x50
+    * 0x50 House of Skulltula
+* Full Trade Quest Shuffle: Scenes 0x60 and 0x62
+    * Items Owned are stored in Death Mountain Trail flags, used with the d-pad in the inventory screen to cycle owned items
+    * Items Traded are stored in Goron City Flags, used to prevent duping trades with the same NPC
+    * Child trade items use bits 0-10, Adult trade items use bits 11-21. See trade_quest_items in ASM/c/trade_quests.c
+    * Owned and Traded flags use the same bit position for the same item (e.g. Weird Egg uses bit 0 for both Owned and Traded flags)
+    * 0x60 Death Mountain Trail
+    * 0x62 Goron City
+* Trade Quest Items Traded: Scene 0x62
```
Comment (fenhl) [inline Notes/extra-save-data.md]: This seems redundant now.

File: ASM/c/get_items.c (hunk 1 above)
Comment (fenhl) [reply ASM/c/get_items.c:429]: Seems fine to me.

**Comment by r0bd0g** [comment]
Is Medi locking you out of Biggy exclusive or not to adult trade shuffle on?

**Comment by mracsys** [comment]
> Is Medi locking you out of Biggy exclusive or not to adult trade shuffle on?

It was not exclusive, and it was fixed for both.

File: Notes/extra-save-data.md (hunk 2 above)
Comment (mracsys) [reply Notes/extra-save-data.md]: You're right, I removed it.

**Comment by mracsys** [comment]
> > Is Medi locking you out of Biggy exclusive or not to adult trade shuffle on?
> 
> It was not exclusive, and it was fixed for both.

I take this back, I don't think it occurred in trade shuffle off seeds for the same reason Biggoron allowed duping. Either way, I tested the behavior with the fix both ways to make sure it worked as expected.

**Comment by r0bd0g** [comment]
And I assume expensive merchants had to be shuffled to have seen any interaction between the two gorons?

**Comment by mracsys** [comment]
> And I assume expensive merchants had to be shuffled to have seen any interaction between the two gorons?

I only tested with them on, but the way Medigoron is shuffled theoretically would cause issues without him shuffled. The root of the issue is in these two lines in the rando item table:

https://github.com/OoTRandomizer/OoT-Randomizer/blob/b1285ac2eea96326f8df7fc86b739d40418c706b/ASM/c/item_table.c#L70

https://github.com/OoTRandomizer/OoT-Randomizer/blob/b1285ac2eea96326f8df7fc86b739d40418c706b/ASM/c/item_table.c#L117

The action IDs (third element) are identical between the Giant's Knife and Biggoron Sword. This ID is used during trade interactions by the NPCs, so it's easier to reference than the item table row ID especially on the ASM side. In the rando `get_item` hook, the trade shuffle flag checks are run based on the incoming vanilla item ID pre-override. Since Medigoron's action ID shows up as the Biggoron Sword when he gives the Giant's Knife (because they're the same item in vanilla, Biggoron sets a separate flag in the save context to prevent the sword from breaking), he should set the Claim Check traded flag before this fix regardless of if he is shuffled. The fix I added was to skip setting the trade flags if the actor providing the item is Medigoron (En_Gm, ID 0x13D). This would not have worked if Medigoron shared the generic Goron actor Biggoron shares with some of the regular-size Gorons (En_Go2, ID 0x1AE).

**Comment by cjohnson57** [comment]
Tysm for the quick turnaround!

**Comment by cjohnson57** [comment]
Seems to be working from my testing

**Pull request merged**